text/0000-custom-dst.md

```diff
@@ -0,0 +1,476 @@
+        ptr: *const (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *const T;
+
+    pub fn from_raw_parts_mut<T: DynamicallySized>(
+        ptr: *mut (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *mut T;
+
+    pub fn metadata<T: DynamicallySized>(
+        ptr: *const T,
+    ) -> <T as DynamicallySized>::Metadata;
+}
+```
+
+and we will introduce two functions into `core::mem`,
+to help people write types with Flexible Array Members:
+
+```rust
+mod core::mem {
+    pub fn size_of_header<T: ?Sized>() -> usize;
+    pub fn align_of_header<T: ?Sized>() -> usize;
+}
+```
+These functions return the size and alignment of the header of a type;
+or, the minimum possible size and alignment, in other words.
+For existing `Sized` types, they are equivalent to `size_of` and `align_of`,
+and for existing DSTs,
+
+```rust
+assert_eq!(size_of_header::<[T]>(), 0);
+assert_eq!(align_of_header::<[T]>(), align_of::<T>());
+assert_eq!(size_of_header::<dyn Trait>(), 0);
+assert_eq!(align_of_header::<dyn Trait>(), 1);
+
+// on 64-bit
+struct RcBox<T: ?Sized> {
+  strong: Cell<usize>,
+  weak: Cell<usize>,
+  value: T,
+}
+assert_eq!(size_of_header::<RcBox<dyn Trait>>(), 16);
+assert_eq!(align_of_header::<RcBox<dyn Trait>>(), 8);
+```
+
+Note that this is a minimum - this means that for `extern type`s,
+they return `0` and `1` respectively.
+
+Notes:
+  - names of the above functions should be bikeshed
+  - `extern type`s do not implement `DynamicallySized`, although in theory one
+    could choose to implement the trait for them
+    (that usecase is not supported by this RFC).
+  - `T: DynamicallySized` bounds imply a `T: ?Sized` bound.
+    - The opposite is not true - `T: ?Sized` implies `T: ?DynamicallySized`
+
+We will also change `CStr` to have the implementation from above.
+
+On an ABI level, we promise that pointers to any type with
+
+```rust
+size_of::<Metadata>() == 0
+&& align_of::<Metadata>() <= align_of::<*const ()>()
+```
+
+are ABI compatible with a C pointer - this is important,
+since we want to be able to write:
+
+```rust
+extern "C" {
+    fn printf(fmt: &CStr, ...) -> c_int;
+}
+```
+
+Unfortunately, we won't be able to change existing declarations in `libc`
+without a new major version.
+
+`as` casts continue to allow
+
+```rust
+fn cast_to_thin<T: DynamicallySized, U: Sized>(t: *const T) -> *const U {
+    t as *const U
+}
+```
+
+so we do not introduce any new functions to access the pointer part
+of the thick pointer.
+
+The `DynamicallySized` trait may be implemented for a struct or union
+-- The author is of the opinion that implementing it for `enum`s is
+unlikely to be useful. That may be a future extension,
+if people are interested
+(once `<T: ?Sized>` is allowed on `enum` declarations).
+
+`DynamicallySized` will be placed into the prelude.
+
+# Drawbacks
+[drawbacks]: #drawback
+
+- More complication in the language.
+- Lack of a `Sized` type dual to these unsized types --
```

Comment:
So that you could define one of these DSTs on the stack, or on the heap, without resorting to unsafe allocator APIs - i.e., `let x = [0, 1, 2]; let y: &[u8] = x;`

Reply:
I think this needs a bit more elaboration in the RFC as well. I'm kind of left wondering why unsized rvalues do not (or could not be extended to) allow this, and also, whether we even actually want these (e.g. accidental usage of dynamically sized types in stack frames could be a performance footgun if the user intended to use a sized type).